### ISSUE TITLE
Relational field custom format via pods_form_ui_field_pick_format_multi(single)_options doesnt provide an autocomplete support

### ISSUE THREAD

**Issue by b0tm1nd** [pods-framework/pods#3897]
Hello.

I am trying to make a custom select2 autocomplete box for relational field type.

1. So at first step I have made:

```
// define the pods_form_ui_field_pick_format_multi_options callback 
function filter_pods_form_ui_field_pick_format_options( $var ) { //var_dump($var);
	$var['autocomplete_plus'] = __( 'Autocomplete Plus', 'pods' );
    // make filter magic happen here... 
    return $var; 
}; 
         
// add the filter 
add_filter( 'pods_form_ui_field_pick_format_multi_options', 'filter_pods_form_ui_field_pick_format_options', 10, 1 ); 
add_filter( 'pods_form_ui_field_pick_format_single_options', 'filter_pods_form_ui_field_pick_format_options', 10, 1 ); 
```

The new type appeared and I have successfully assigned it to my field.

2. Then I done:

```
// define the pods_form_ui_field_pick_input callback 
function action_pods_form_ui_field_pick_input_autocomplete_plus( $name, $value, $options, $pod, $id ) { 
    $f = (file_get_contents(dirname(__FILE__).'/../pods/ui/fields/select2.php'));
    //$f = str_replace('minimumInputLength : 1', 'minimumInputLength : 0', $f);
    //$f = str_replace('data : function ( term, page ) {', "data : function ( term, page ) { if (!term) term = '%';", $f);
    eval('?>' . $f );
}; 
         
// add the action 
add_action( 'pods_form_ui_field_pick_input_multi_autocomplete_plus', 'action_pods_form_ui_field_pick_input_autocomplete_plus', 10, 6 ); 
add_action( 'pods_form_ui_field_pick_input_single_autocomplete_plus', 'action_pods_form_ui_field_pick_input_autocomplete_plus', 10, 6 ); 
```

Here I commented out my changes of select2.php template. So I just show it unmodified.

**Comment by b0tm1nd** [comment]
I am progressing with this. After having disabled all other pods related code, it started working. But now I am having all field values in $options['data'] array once using the custom type! While it should contain only the chosen values.

**Comment by b0tm1nd** [comment]
Also I have noticed that its only possible to have a single select this way. By digging select2.php template, I figured out that $form_field_type variable is not set. So doing like:

```
// define the pods_form_ui_field_pick_input callback 
function action_pods_form_ui_field_pick_input_autocomplete_plus( $name, $value, $options, $pod, $id ) { 
    $form_field_type = 'pick'; 
    $f = (file_get_contents(dirname(__FILE__).'/../pods/ui/fields/select2.php'));
    //$f = str_replace('minimumInputLength : 1', 'minimumInputLength : 0', $f);
    //$f = str_replace('data : function ( term, page ) {', "data : function ( term, page ) { if (!term) term = '%';", $f);
    eval('?>' . $f );
}; 
         
// add the action 
add_action( 'pods_form_ui_field_pick_input_multi_autocomplete_plus', 'action_pods_form_ui_field_pick_input_autocomplete_plus', 10, 6 ); 
add_action( 'pods_form_ui_field_pick_input_single_autocomplete_plus', 'action_pods_form_ui_field_pick_input_autocomplete_plus', 10, 6 ); 
```

Will make it visually work perfect. Except still having $options['data'] full of values overloads page HTML/JS with data when having thousands of  records.

**Comment by pglewis** [comment]
As a heads-up: 2.7 basically jettisons the majority if not all of our existing Javascript and the template related to select2 and updates to v4 of select2.  No more mixed PHP and Javascript.  You may want to setup a test environment with the 2.7 branch to help expose any issues before we go into beta soon.

**Comment by b0tm1nd** [comment]
Ok I give up doing this without modifying the core.
Apart from $form_field_type = 'pick'; I had to also add  $ajax = true; so ajax will be enabled on edit.

Now about core modification. It ended up that when you create a custom pick format you are not able to use autocomplete features! Ie you are not allowed to only only a part of data!
Plus, contrary, when you are in autocomplete mode, you are not allowed to get results without a query!

My whole idea was to show a list of results just by clicking a field and also allowing to search. So this is to be achieved by modifying the next lines in pick.php:

Around line 1498 change:

```
                 if ( 'single' == pods_var( self::$type . '_format_type', $options, 'single' ) && 'autocomplete' == pods_var( self::$type . '_format_single', $options, 'dropdown' ) )
                    $autocomplete = true;
                elseif ( 'multi' == pods_var( self::$type . '_format_type', $options, 'single' ) && 'autocomplete' == pods_var( self::$type . '_format_multi', $options, 'checkbox' ) )
                    $autocomplete = true;   
```

To:

```
                if ( 'single' == pods_var( self::$type . '_format_type', $options, 'single' ) && strstr(pods_var( self::$type . '_format_single', $options, 'dropdown' ), 'autocomplete') )
                    $autocomplete = true;
                elseif ( 'multi' == pods_var( self::$type . '_format_type', $options, 'single' ) && strstr(pods_var( self::$type . '_format_multi', $options, 'checkbox' ), 'autocomplete' ))
                    $autocomplete = true;
```

Maybe not the best solution, but the simpliest. If a custom pick format will contain 'autocomplete' word, autocomplete mode will get triggered.

Next, around line 1792 comment out like this:

```
        /*if ( !isset( $params->query ) || strlen( trim( $params->query ) ) < 1 )
            pods_error( __( 'Invalid field request', 'pods' ), PodsInit::$admin );
        else*/
```

Which will allow empty queries.

Any chances these changes to be applied?

**Comment by b0tm1nd** [comment]
pglewis, thanks, but the limitation lies in pick.php, see my above comment. The issues are allowing autocomplete mode for custom pick formats plus empty queries in autocomplete mode.

**Comment by b0tm1nd** [comment]
Maybe empty queries in autocomplete mode might be achieved with pods_form_ui_field_pick_autocomplete_lookup, havent checked yet, but not allowing autocomplete for a custom pick format seems like a bug to me.

**Comment by pglewis** [comment]
pick.php is a big part of overhaul in 2.7 (it's a mess there right now, in fact, as things are transitioning).  I have not had time to digest all the information here, I have my hands full in other areas right now to get 2.7 to a beta stage, but we'll definitely want to make sure we support everything properly.  Pick fields in 2.7 don't even resemble what we're doing in the 2.x branch, so it'd be helpful for both of us if you have time to experiment some on that branch in a non-production environment.

**Comment by b0tm1nd** [comment]
Hmm. Not sure what you mean with "don't even resemble" (I'm not a native English speaker). I've checked https://github.com/pods-framework/pods/blob/release/2.7/classes/fields/pick.php and this autocomplete limitation is still there..

**Comment by pglewis** [comment]
To clarify the meaning: how we implement pick fields in 2.7 has completely changed.  It would be wasteful to dig deeper into where the problem is in the 2.x branch since 2.7 will soon replace it.  This is a lot of detail to read and digest at this moment, so I can't address your issue right this moment, but if it is broken we will have it working for 2.7.

**Comment by b0tm1nd** [comment]
Ok thanks for clarifying! But I assume its still there if those lines of code are still present which define the autocomplete mode.

And about 2.7 version, how smooth the transition will be? I am now building a project with 2.x version downloaded from WP site.

**Comment by pglewis** [comment]
The transition should smooth in the sense that none of the changes should break backward compatibility.  I'll eventually get more time to look at this issue more closely.

**Comment by b0tm1nd** [comment]
Sounds good!
And thanks. This is very simple actually. I am trying to achieve this:
![screenshot from 2016-12-15 05 27 29](https://cloud.githubusercontent.com/assets/3400283/21210992/46776afe-c288-11e6-863f-34eeb981ef3a.png)
You click and instantly see the results but also able to search.
And done this via adding a new pick format using pods_form_ui_field_pick_format_multi_options action so in Pods admin it looks like this:
![screenshot from 2016-12-15 05 34 56](https://cloud.githubusercontent.com/assets/3400283/21211033/9eaad7ce-c288-11e6-98cc-9e3eeb8c5010.png)

**Comment by quasel** [comment]
maybe a nice addition to 2.8?

**Comment by sc0ttkclark** [comment]
Following up on this, while we do want to continue to improve our hooks so more things are customizable -- the need here is now solved by the options / UI in Pods 2.8.
